caikit/core/model_management/local_model_trainer.py

```diff
@@ -0,0 +1,328 @@
+                result = caikit.load(self.save_path)
+            else:
+                result = self._worker.get_or_throw()
+            return result
+
+        ## Impl ##
+
+        def _train_and_save(self, *args, return_completion_time=False, **kwargs):
+            """Function that will run in the worker thread"""
+            with alog.ContextTimer(log.debug, "Training %s finished in: ", self.id):
+                trained_model = self._module_class.train(*args, **kwargs)
+            if self.save_path is not None:
+                log.debug("Saving training %s to %s", self.id, self.save_path)
+                with alog.ContextTimer(log.debug, "Training %s saved in: ", self.id):
+                    trained_model.save(self.save_path)
+            self._completion_time = datetime.now()
+            log.debug2("Completion time for %s: %s", self.id, self._completion_time)
+            if return_completion_time:
+                return self._completion_time
+            return trained_model
+
+    ## Interface ##
+
+    # Expression for parsing retention policy
+    _timedelta_expr = re.compile(
+        r"^((?P<days>\d+?)d)?((?P<hours>\d+?)h)?((?P<minutes>\d+?)m)?((?P<seconds>\d*\.?\d*?)s)?$"
+    )
+
+    def __init__(self, config: aconfig.Config, instance_name: str):
+        """Initialize with a shared dict of all trainings"""
+        self._instance_name = instance_name
+        self._use_subprocess = config.get("use_subprocess", False)
+        self._retention_duration = config.get("retention_duration")
+        if self._retention_duration is not None:
+            try:
+                log.debug2("Parsing retention duration: %s", self._retention_duration)
+                self._retention_duration = timedelta(
+                    **{
+                        key: float(val)
+                        for key, val in self._timedelta_expr.match(
+                            self._retention_duration
+                        )
+                        .groupdict()
+                        .items()
+                        if val is not None
+                    }
+                )
+            except AttributeError:
+                error(
+                    "<COR63897671E>",
+                    ValueError(
+                        f"Invalid retention_duration: {self._retention_duration}"
+                    ),
+                )
+
+        # The shared dict of futures and a lock to serialize mutations to it
+        self._futures = {}
+        self._futures_lock = threading.Lock()
+
+    def train(
+        self,
+        module_class: Type[ModuleBase],
+        *args,
+        save_path: Optional[str] = None,
+        save_with_id: bool = False,
+        external_training_id: Optional[str] = None,
+        **kwargs,
+    ) -> "LocalModelFuture":
+        """Start training the given module and return a future to the trained
+        model instance
+        """
+        # Always purge old futures
+        self._purge_old_futures()
```

Comment:
Why do these need purging at train time?

Reply:
Hm, good question. I think I threw that in there just to say "any time the map is touched, purge," but I think you're right that this should be removed

Reply:
Actually, I think the reason I did it here was so that we could guarantee that the in-memory dict wouldn't grow without bound. This is all pretty far-fetched since this implementation won't be used in a production environment anyway, but if we only purge on lookup, we could theoretically just keep submitting training jobs and having the map keep growing.